File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
         int[] left = new int[height.length];
         int[] right = new int[height.length];
         int result = 0;
         int max=-1;
         for(int i=0;i<height.length;i++){
            if(height[i]>max){
                max=height[i];
            }
            left[i]=max;   
         }
         max=-1;
         for(int i=height.length-1;i>=0;i--){
            if(height[i]>max){
                max=height[i];
            }
            right[i]=max;
         }
         for(int i=0;i<height.length;i++){
            result=result+Math.min(left[i],right[i])-height[i];
         }
         return result;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        int result=0;
        vector<int>lm(n);
        vector<int>rm(n);
        
        lm[0]=height[0];
        for(int i=1;i<n;i++){
            lm[i]=max(lm[i-1],height[i]);
        }
        rm[n-1]=height[n-1];
        for(int i=n-2;i>=0;i--){
            rm[i]=max(rm[i+1],height[i]);
        }
        for(int i=0;i<n;i++){
        result = result+min(lm[i],rm[i])-height[i];
    }
    return result;
    }
};